<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head><title>R: Tools for curating, analyzing, and manipulating biological
sequences</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>
<h1> Tools for curating, analyzing, and manipulating biological
sequences
<img class="toplogo" src="../../../doc/html/Rlogo.svg" alt="[R logo]" />
</h1>
<hr/>
<div style="text-align: center;">
<a href="../../../doc/html/packages.html"><img class="arrow" src="../../../doc/html/left.jpg" alt="[Up]" /></a>
<a href="../../../doc/html/index.html"><img class="arrow" src="../../../doc/html/up.jpg" alt="[Top]" /></a>
</div><h2>Documentation for package &lsquo;DECIPHER&rsquo; version 2.18.1</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
<li><a href="../doc/index.html">User guides, package vignettes and other documentation.</a></li>
</ul>

<h2>Help Pages</h2>


<table width="100%">
<tr><td style="width: 25%;"><a href="DECIPHER-package.html">DECIPHER-package</a></td>
<td>Tools for curating, analyzing, and manipulating biological sequences</td></tr>
<tr><td style="width: 25%;"><a href="AA_REDUCED.html">AA_REDUCED</a></td>
<td>Reduced amino acid alphabets</td></tr>
<tr><td style="width: 25%;"><a href="Add2DB.html">Add2DB</a></td>
<td>Add Data to a Database</td></tr>
<tr><td style="width: 25%;"><a href="AdjustAlignment.html">AdjustAlignment</a></td>
<td>Improve An Existing Alignment By Adjusting Gap Placements</td></tr>
<tr><td style="width: 25%;"><a href="AlignDB.html">AlignDB</a></td>
<td>Align Two Sets of Aligned Sequences in a Sequence Database</td></tr>
<tr><td style="width: 25%;"><a href="AlignProfiles.html">AlignProfiles</a></td>
<td>Align Two Sets of Aligned Sequences</td></tr>
<tr><td style="width: 25%;"><a href="AlignSeqs.html">AlignSeqs</a></td>
<td>Align a Set of Unaligned Sequences</td></tr>
<tr><td style="width: 25%;"><a href="AlignSynteny.html">AlignSynteny</a></td>
<td>Pairwise Aligns Syntenic Blocks</td></tr>
<tr><td style="width: 25%;"><a href="AlignTranslation.html">AlignTranslation</a></td>
<td>Align Sequences By Their Amino Acid Translation</td></tr>
<tr><td style="width: 25%;"><a href="AmplifyDNA.html">AmplifyDNA</a></td>
<td>Simulate Amplification of DNA by PCR</td></tr>
<tr><td style="width: 25%;"><a href="Array2Matrix.html">Array2Matrix</a></td>
<td>Create a Matrix Representation of a Microarray</td></tr>
<tr><td style="width: 25%;"><a href="BrowseDB.html">BrowseDB</a></td>
<td>View a Database Table in a Web Browser</td></tr>
<tr><td style="width: 25%;"><a href="BrowseSeqs.html">BrowseSeqs</a></td>
<td>View Sequences in a Web Browser</td></tr>
<tr><td style="width: 25%;"><a href="Taxa-class.html">c.Taxa</a></td>
<td>Taxa training and testing objects</td></tr>
<tr><td style="width: 25%;"><a href="CalculateEfficiencyArray.html">CalculateEfficiencyArray</a></td>
<td>Predict the Hybridization Efficiency of Probe/Target Sequence Pairs</td></tr>
<tr><td style="width: 25%;"><a href="CalculateEfficiencyFISH.html">CalculateEfficiencyFISH</a></td>
<td>Predict Thermodynamic Parameters of Probe/Target Sequence Pairs</td></tr>
<tr><td style="width: 25%;"><a href="CalculateEfficiencyPCR.html">CalculateEfficiencyPCR</a></td>
<td>Predict Amplification Efficiency of Primer Sequences</td></tr>
<tr><td style="width: 25%;"><a href="Codec.html">Codec</a></td>
<td>Compression/Decompression of Character Vectors</td></tr>
<tr><td style="width: 25%;"><a href="ConsensusSequence.html">ConsensusSequence</a></td>
<td>Create a Consensus Sequence</td></tr>
<tr><td style="width: 25%;"><a href="Cophenetic.html">Cophenetic</a></td>
<td>Compute cophenetic distances on dendrogram objects</td></tr>
<tr><td style="width: 25%;"><a href="CorrectFrameshifts.html">CorrectFrameshifts</a></td>
<td>Corrects Frameshift Errors In Protein Coding Sequences</td></tr>
<tr><td style="width: 25%;"><a href="CreateChimeras.html">CreateChimeras</a></td>
<td>Create Artificial Chimeras</td></tr>
<tr><td style="width: 25%;"><a href="DB2Seqs.html">DB2Seqs</a></td>
<td>Export Database Sequences to a FASTA or FASTQ File</td></tr>
<tr><td style="width: 25%;"><a href="DECIPHER-package.html">DECIPHER</a></td>
<td>Tools for curating, analyzing, and manipulating biological sequences</td></tr>
<tr><td style="width: 25%;"><a href="deltaGrules.html">deltaGrules</a></td>
<td>Free Energy of Hybridization of Probe/Target Quadruplets on Microarrays</td></tr>
<tr><td style="width: 25%;"><a href="deltaHrules.html">deltaHrules</a></td>
<td>Change in Enthalpy of Hybridization of DNA/DNA Quadruplets in Solution</td></tr>
<tr><td style="width: 25%;"><a href="deltaHrulesRNA.html">deltaHrulesRNA</a></td>
<td>Change in Enthalpy of Hybridization of RNA/RNA Quadruplets in Solution</td></tr>
<tr><td style="width: 25%;"><a href="deltaSrules.html">deltaSrules</a></td>
<td>Change in Entropy of Hybridization of DNA/DNA Quadruplets in Solution</td></tr>
<tr><td style="width: 25%;"><a href="deltaSrulesRNA.html">deltaSrulesRNA</a></td>
<td>Change in Entropy of Hybridization of RNA/RNA Quadruplets in Solution</td></tr>
<tr><td style="width: 25%;"><a href="DesignArray.html">DesignArray</a></td>
<td>Design a Set of DNA Microarray Probes for Detecting Sequences</td></tr>
<tr><td style="width: 25%;"><a href="DesignPrimers.html">DesignPrimers</a></td>
<td>Design PCR Primers Targeting a Specific Group of Sequences</td></tr>
<tr><td style="width: 25%;"><a href="DesignProbes.html">DesignProbes</a></td>
<td>Design FISH Probes Targeting a Specific Group of Sequences</td></tr>
<tr><td style="width: 25%;"><a href="DesignSignatures.html">DesignSignatures</a></td>
<td>Design PCR Primers for Amplifying Group-Specific Signatures</td></tr>
<tr><td style="width: 25%;"><a href="DigestDNA.html">DigestDNA</a></td>
<td>Simulate Restriction Digestion of DNA</td></tr>
<tr><td style="width: 25%;"><a href="Disambiguate.html">Disambiguate</a></td>
<td>Expand Ambiguities into All Permutations of a DNAStringSet</td></tr>
<tr><td style="width: 25%;"><a href="DistanceMatrix.html">DistanceMatrix</a></td>
<td>Calculate the Distances Between Sequences</td></tr>
<tr><td style="width: 25%;"><a href="ExtractGenes.html">ExtractGenes</a></td>
<td>Extract Predicted Genes from a Genome</td></tr>
<tr><td style="width: 25%;"><a href="FindChimeras.html">FindChimeras</a></td>
<td>Find Chimeras in a Sequence Database</td></tr>
<tr><td style="width: 25%;"><a href="FindGenes.html">FindGenes</a></td>
<td>Find Genes in a Genome</td></tr>
<tr><td style="width: 25%;"><a href="FindSynteny.html">FindSynteny</a></td>
<td>Finds Synteny in a Sequence Database</td></tr>
<tr><td style="width: 25%;"><a href="FormGroups.html">FormGroups</a></td>
<td>Forms Groups By Rank</td></tr>
<tr><td style="width: 25%;"><a href="Genes-class.html">Genes-class</a></td>
<td>Genes objects and accessors</td></tr>
<tr><td style="width: 25%;"><a href="HEC_MI.html">HEC_MI1</a></td>
<td>Mutual Information for Protein Secondary Structure Prediction</td></tr>
<tr><td style="width: 25%;"><a href="HEC_MI.html">HEC_MI2</a></td>
<td>Mutual Information for Protein Secondary Structure Prediction</td></tr>
<tr><td style="width: 25%;"><a href="IdClusters.html">IdClusters</a></td>
<td>Cluster Sequences By Distance or Sequence</td></tr>
<tr><td style="width: 25%;"><a href="IdConsensus.html">IdConsensus</a></td>
<td>Create Consensus Sequences by Groups</td></tr>
<tr><td style="width: 25%;"><a href="IdentifyByRank.html">IdentifyByRank</a></td>
<td>Identify By Taxonomic Rank</td></tr>
<tr><td style="width: 25%;"><a href="IdLengths.html">IdLengths</a></td>
<td>Determine the Number of Bases, Nonbases, and Width of Each Sequence</td></tr>
<tr><td style="width: 25%;"><a href="IdTaxa.html">IdTaxa</a></td>
<td>Assign Sequences a Taxonomic Classification</td></tr>
<tr><td style="width: 25%;"><a href="LearnTaxa.html">LearnTaxa</a></td>
<td>Train a Classifier for Assigning Taxonomy</td></tr>
<tr><td style="width: 25%;"><a href="MapCharacters.html">MapCharacters</a></td>
<td>Map Changes in Ancestral Character States</td></tr>
<tr><td style="width: 25%;"><a href="MaskAlignment.html">MaskAlignment</a></td>
<td>Mask Highly Variable Regions of An Alignment</td></tr>
<tr><td style="width: 25%;"><a href="MeltDNA.html">MeltDNA</a></td>
<td>Simulate Melting of DNA</td></tr>
<tr><td style="width: 25%;"><a href="MIQS.html">MIQS</a></td>
<td>MIQS Amino Acid Substitution Matrix</td></tr>
<tr><td style="width: 25%;"><a href="MODELS.html">MODELS</a></td>
<td>Available Models of DNA Evolution</td></tr>
<tr><td style="width: 25%;"><a href="NNLS.html">NNLS</a></td>
<td>Sequential Coordinate-wise Algorithm for the Non-negative Least Squares Problem</td></tr>
<tr><td style="width: 25%;"><a href="OrientNucleotides.html">OrientNucleotides</a></td>
<td>Orient Nucleotide Sequences</td></tr>
<tr><td style="width: 25%;"><a href="Synteny-class.html">pairs.Synteny</a></td>
<td>Synteny blocks and hits</td></tr>
<tr><td style="width: 25%;"><a href="PFASUM.html">PFASUM</a></td>
<td>PFASUM Amino Acid Substitution Matrices</td></tr>
<tr><td style="width: 25%;"><a href="Genes-class.html">plot.Genes</a></td>
<td>Genes objects and accessors</td></tr>
<tr><td style="width: 25%;"><a href="Synteny-class.html">plot.Synteny</a></td>
<td>Synteny blocks and hits</td></tr>
<tr><td style="width: 25%;"><a href="Taxa-class.html">plot.Taxa</a></td>
<td>Taxa training and testing objects</td></tr>
<tr><td style="width: 25%;"><a href="PredictDBN.html">PredictDBN</a></td>
<td>Predict RNA Secondary Structure in Dot-Bracket Notation</td></tr>
<tr><td style="width: 25%;"><a href="PredictHEC.html">PredictHEC</a></td>
<td>Predict Protein Secondary Structure as Helix, Beta-Sheet, or Coil</td></tr>
<tr><td style="width: 25%;"><a href="Genes-class.html">print.Genes</a></td>
<td>Genes objects and accessors</td></tr>
<tr><td style="width: 25%;"><a href="Synteny-class.html">print.Synteny</a></td>
<td>Synteny blocks and hits</td></tr>
<tr><td style="width: 25%;"><a href="Taxa-class.html">print.Taxa</a></td>
<td>Taxa training and testing objects</td></tr>
<tr><td style="width: 25%;"><a href="ReadDendrogram.html">ReadDendrogram</a></td>
<td>Read a Dendrogram from a Newick Formatted File</td></tr>
<tr><td style="width: 25%;"><a href="RemoveGaps.html">RemoveGaps</a></td>
<td>Remove Gap Characters in Sequences</td></tr>
<tr><td style="width: 25%;"><a href="RESTRICTION_ENZYMES.html">RESTRICTION_ENZYMES</a></td>
<td>Common Restriction Enzyme's Cut Sites</td></tr>
<tr><td style="width: 25%;"><a href="SearchDB.html">SearchDB</a></td>
<td>Obtain Specific Sequences from a Database</td></tr>
<tr><td style="width: 25%;"><a href="Seqs2DB.html">Seqs2DB</a></td>
<td>Add Sequences from Text File to Database</td></tr>
<tr><td style="width: 25%;"><a href="StaggerAlignment.html">StaggerAlignment</a></td>
<td>Produce a Staggered Alignment</td></tr>
<tr><td style="width: 25%;"><a href="Synteny-class.html">Synteny-class</a></td>
<td>Synteny blocks and hits</td></tr>
<tr><td style="width: 25%;"><a href="Taxa-class.html">Taxa-class</a></td>
<td>Taxa training and testing objects</td></tr>
<tr><td style="width: 25%;"><a href="TerminalChar.html">TerminalChar</a></td>
<td>Determine the Number of Terminal Characters</td></tr>
<tr><td style="width: 25%;"><a href="TileSeqs.html">TileSeqs</a></td>
<td>Form a Set of Tiles for Each Group of Sequences.</td></tr>
<tr><td style="width: 25%;"><a href="TrainingSet_16S.html">TrainingSet_16S</a></td>
<td>Training Set for Classification of 16S rRNA Gene Sequences</td></tr>
<tr><td style="width: 25%;"><a href="TrimDNA.html">TrimDNA</a></td>
<td>Trims DNA Sequences to the High Quality Region Between Patterns</td></tr>
<tr><td style="width: 25%;"><a href="WriteDendrogram.html">WriteDendrogram</a></td>
<td>Write a Dendrogram to Newick Format</td></tr>
<tr><td style="width: 25%;"><a href="WriteGenes.html">WriteGenes</a></td>
<td>Write Genes to a File</td></tr>
<tr><td style="width: 25%;"><a href="Genes-class.html">[.Genes</a></td>
<td>Genes objects and accessors</td></tr>
<tr><td style="width: 25%;"><a href="Synteny-class.html">[.Synteny</a></td>
<td>Synteny blocks and hits</td></tr>
<tr><td style="width: 25%;"><a href="Taxa-class.html">[.Taxa</a></td>
<td>Taxa training and testing objects</td></tr>
</table>
</body></html>
